### ISSUE TITLE
ERROR: HDF5OutputLayer: output file 'C:\Users\....tmp' already exists

### ISSUE THREAD

**Issue by wolfego** [pluskid/Mocha.jl#58]
Below is console output describing error, I presume this is wrong place to ask for help but would welcome any suggestion as to where to start / look for help?

21-Mar 16:11:02:INFO:root:Configuring Mocha...
21-Mar 16:11:02:INFO:root: * CUDA       disabled by default
21-Mar 16:11:02:INFO:root: * Native Ext disabled by default
21-Mar 16:11:02:INFO:root:Mocha configured, continue loading module...
-- Testing network topology with duplicated blobs
21-Mar 16:11:08:INFO:root:Constructing net net on CPUBackend...
21-Mar 16:11:08:INFO:root:Topological sorting 1 layers...
21-Mar 16:11:08:INFO:root:Constructing net net on CPUBackend...
21-Mar 16:11:08:INFO:root:Topological sorting 2 layers...
-- Testing network topology with missing blobs
21-Mar 16:11:08:INFO:root:Constructing net net on CPUBackend...
21-Mar 16:11:08:INFO:root:Topological sorting 1 layers...
-- Testing network topology with circular dependency
21-Mar 16:11:08:INFO:root:Constructing net net on CPUBackend...
21-Mar 16:11:08:INFO:root:Topological sorting 2 layers...
-- Testing network topology with multiple back-propagate path
    > Good blob sharing
21-Mar 16:11:08:INFO:root:Constructing net net on CPUBackend...
21-Mar 16:11:08:INFO:root:Topological sorting 5 layers...
21-Mar 16:11:08:INFO:root:Setup layers...
21-Mar 16:11:09:INFO:root:Network constructed!
21-Mar 16:11:09:DEBUG:root:Destroying network net
    > Bad blob sharing
21-Mar 16:11:09:INFO:root:Constructing net net on CPUBackend...
21-Mar 16:11:09:INFO:root:Topological sorting 6 layers...
21-Mar 16:11:09:INFO:root:Setup layers...
21-Mar 16:11:09:INFO:root:Network constructed!
-- Testing network topology with dangling blob
    > Good case
21-Mar 16:11:09:INFO:root:Constructing net net on CPUBackend...
21-Mar 16:11:09:INFO:root:Topological sorting 4 layers...
21-Mar 16:11:09:INFO:root:Setup layers...
ERROR: HDF5OutputLayer: output file 'C:\Users\wolfe_000\AppData\Local\Temp\jul4BB6.tmp' already exists
 in error at error.jl:21 (repeats 2 times)
while loading C:\Users\wolfe_000\.julia\v0.3\Mocha\test\net/topology.jl, in expression starting on line 112
while loading C:\Users\wolfe_000\.julia\v0.3\Mocha\test\runtests.jl, in expression starting on line 33

================================[ ERROR: Mocha ]================================

failed process: Process(`'c:\Program Files (x86)\Julia\resources\app\julia\bin/julia' 'C:\Users\wolfe_000\.julia\v0.3\Mocha\test\runtests.jl'`, ProcessExited(1)) [1]

**Comment by pluskid** [comment]
@wolfego I'm not sure what happens to Windows. It seems like tmp files generated during unit-test are not successfully deleted. I modified the unit test slightly. Could you test whether the master fixed the problem?

**Comment by wolfego** [comment]
I tried the two versions you created over last 24 hours, neither worked for me.  I also tried something I read here https://github.com/JuliaLang/julia/issues/9053 due to an apparently known Julia issue, specifically:

function tempname()
    (f, io) = mktemp()
    close(io)
    f
end

This did not work for me either.

I commented out these lines in topology.jl to get through to the rest of the Mocha tests,

#test_net_topology_multiple_bp(backend)
#test_net_topology_dangling_blob(backend)

Then had a similar issue in hdf5-data.jl

Commented these lines out in hdf5-data.jl starting at line 130 to get through to the rest of the Mocha tests,

 #test_hdf5_data_layer(backend, Float32, 1e-3)
 #test_hdf5_data_layer(backend, Float64, 1e-7)
 #test_hdf5_data_layer_shuffle(backend, Float32)
 #test_hdf5_data_layer_shuffle(backend, Float64)

This allowed me to get through all the remaining tests.

In summary it seems tempfile() does not behave the same in Windows as *nix and is a known problem but I am not sufficiently expert to work around it.

**Comment by wolfego** [comment]
Sorry, I should have said tempname() instead of tempfile() in last sentence of previous post

**Comment by pluskid** [comment]
@wolfego Thanks for the pointer to Julia issue. I made a workaround in the latest commit by using `tempdir` and manually generate a random string to create tmp file instead of using the problematic `tempname` function. I hope it will work now on Windows. Could you please run the test on your environment again? Thank you very much!

**Comment by wolfego** [comment]
Thanks much!!! Made it through topology.jl, all good.  Had error in hdf5-data.jl as listed below:

-- Testing HDF5 Output Layer on CPUBackend{Float32}...
    > (5,8,2)
23-Mar 22:34:48:WARNING:root:HDF5OutputLayer: output file 'C:\Users\wolfe_000\AppData\Local\Temp\Mocha-528-GYeLx0Efix1sQQQ4Fy17c4jSftq3PEOk.hdf5' already exists, overwriting
-- Testing HDF5 Output Layer on CPUBackend{Float64}...
    > (2,4,1,7,4)
23-Mar 22:34:48:WARNING:root:HDF5OutputLayer: output file 'C:\Users\wolfe_000\AppData\Local\Temp\Mocha-528-jVNc40sYHVaM6NGsmMrFenzEXKGQzfOM.hdf5' already exists, overwriting
-- Testing  HDF5 Data Layer on CPUBackend{Float32}...
    > (3,)
-- Testing (Async) HDF5 Data Layer on CPUBackend{Float32}...
    > (1,6,2,5,3,3)
23-Mar 22:34:50:INFO:root:AsyncHDF5DataLayer: Stopping IO task...
23-Mar 22:34:50:INFO:root:AsyncHDF5DataLayer: IO Task reaching the end...
ERROR: unlink: resource busy or locked (EBUSY)
 in unlink at fs.jl:97 (repeats 2 times)
while loading C:\Users\wolfe_000\.julia\v0.3\Mocha\test\layers/hdf5-data.jl, in expression starting on line 137
while loading C:\Users\wolfe_000\.julia\v0.3\Mocha\test\runtests.jl, in expression starting on line 77

**Comment by pluskid** [comment]
@wolfego Thanks! Sorry for making this so tedious. I just pushed another workaround for the HDF5 layer. There seems to be problem deleting a file right after it was closed (in another Julia coroutine) under Windows. I wrap the `rm` in a try-catch to stop it from crashing the whole unit-test. Could you have a test again with it? Thank you very much!

**Comment by wolfego** [comment]
Got a little further this time and again thanks for the help!  Made it to the two data layer shuffle tests but neither of those will run.  Here is the output:

-- Testing DataTransformers on CPUBackend{Float32}...
    > SubMean
    > Scale
-- Testing DataTransformers on CPUBackend{Float64}...
    > SubMean
    > Scale
-- Testing  HDF5 Data Layer on CPUBackend{Float32}...
    > (1,4)
-- Testing (Async) HDF5 Data Layer on CPUBackend{Float32}...
    > (3,2,4)
24-Mar 21:40:31:INFO:root:AsyncHDF5DataLayer: Stopping IO task...
24-Mar 21:40:31:INFO:root:AsyncHDF5DataLayer: IO Task reaching the end...
unlink: resource busy or locked (EBUSY)
unlink: resource busy or locked (EBUSY)
unlink: resource busy or locked (EBUSY)
-- Testing  HDF5 Data Layer on CPUBackend{Float64}...
    > (6,1,5,8,2)
-- Testing (Async) HDF5 Data Layer on CPUBackend{Float64}...
    > (8,)
24-Mar 21:40:32:INFO:root:AsyncHDF5DataLayer: Stopping IO task...
24-Mar 21:40:32:INFO:root:AsyncHDF5DataLayer: IO Task reaching the end...
unlink: resource busy or locked (EBUSY)
unlink: resource busy or locked (EBUSY)
unlink: resource busy or locked (EBUSY)
-- Testing  HDF5 Data Layer (shuffle,n=6,b=4) on CPUBackend{Float32}...
ERROR: could not create file mapping: Access is denied. 
 in error at error.jl:21 (repeats 2 times)
while loading C:\Users\wolfe_000\.julia\v0.3\Mocha\test\layers/hdf5-data.jl, in expression starting on line 142
while loading C:\Users\wolfe_000\.julia\v0.3\Mocha\test\runtests.jl, in expression starting on line 77

================================[ ERROR: Mocha ]================================

failed process: Process(`'c:\Program Files (x86)\Julia\resources\app\julia\bin/julia' 'C:\Users\wolfe_000\.julia\v0.3\Mocha\test\runtests.jl'`, ProcessExited(1)) [1]

**Comment by wolfego** [comment]
Thanks for the reliable help and apologies I can't debug more from this
end, your latest update got us a little further down the road, I posted
results on git.

**Comment by pluskid** [comment]
@wolfego, great! At this point I acutally know what is going on. The HDF5 men map is known to not work for windows. I should disable unit test on shuffled hdf5 under windows. I'll do that soon!

**Comment by pluskid** [comment]
@wolfego I updated the code again, this time the HDF5-data layer with shuffling should be disabled for unit-test under Windows, as that is known to not function properly.

That being said, `HDF5DataLayer` itself is sort of legacy now, as `AsyncHDF5DataLayer` is recommended for almost all the cases. And `AsyncHDF5DataLayer` has a better implementation for shuffling that does not require `memmap`, and should function proper also under Windows.

**Comment by wolfego** [comment]
@pluskid Thanks again, now only have some issue with cleanup at end of test

-- Testing  HDF5 Data Layer on CPUBackend{Float32}...
    > (3,)
-- Testing (Async) HDF5 Data Layer on CPUBackend{Float32}...
    > (1,6,2,5,3,3)
25-Mar 22:23:19:INFO:root:AsyncHDF5DataLayer: Stopping IO task...
25-Mar 22:23:19:INFO:root:AsyncHDF5DataLayer: IO Task reaching the end...
unlink: resource busy or locked (EBUSY)
unlink: resource busy or locked (EBUSY)
unlink: resource busy or locked (EBUSY)
-- Testing  HDF5 Data Layer on CPUBackend{Float64}...
    > (2,8,7)
-- Testing (Async) HDF5 Data Layer on CPUBackend{Float64}...
    > (5,)
25-Mar 22:23:19:INFO:root:AsyncHDF5DataLayer: Stopping IO task...
25-Mar 22:23:19:INFO:root:AsyncHDF5DataLayer: IO Task reaching the end...
unlink: resource busy or locked (EBUSY)
unlink: resource busy or locked (EBUSY)
unlink: resource busy or locked (EBUSY)
-- Testing (Async) HDF5 Data Layer (shuffle,n=6,b=4) on CPUBackend{Float32}...
25-Mar 22:23:19:INFO:root:AsyncHDF5DataLayer: Stopping IO task...
25-Mar 22:23:19:INFO:root:AsyncHDF5DataLayer: IO Task reaching the end...
ERROR: unlink: resource busy or locked (EBUSY)
 in unlink at fs.jl:97 (repeats 2 times)
while loading C:\Users\wolfe_000\.julia\v0.3\Mocha\test\layers/hdf5-data.jl, in expression starting on line 145
while loading C:\Users\wolfe_000\.julia\v0.3\Mocha\test\runtests.jl, in expression starting on line 77

================================[ ERROR: Mocha ]================================

failed process: Process(`'c:\Program Files (x86)\Julia\resources\app\julia\bin/julia' 'C:\Users\wolfe_000\.julia\v0.3\Mocha\test\runtests.jl'`, ProcessExited(1)) [1]

**Comment by pluskid** [comment]
@wolfego Thanks very much for the patient! Everything should be fine now!

**Comment by wolfego** [comment]
@pluskid Thanks much!  If I can assist testing on Windows platform in future please let me know.
INFO: Mocha tests passed

**Comment by pluskid** [comment]
@wolfego Great! :D
